stores/metadata.go

```diff
@@ -2151,79 +2171,125 @@ func (s *SQLStore) UnhealthySlabs(ctx context.Context, healthCutoff float64, set
-				return fmt.Errorf("failed to fetch slab: %w", err)
-			}
+	}
+
+	// create slabs that don't exist yet
+	err := tx.
+		Clauses(clause.OnConflict{
+			DoNothing: true,
+			Columns:   []clause.Column{{Name: "key"}},
+		}).
+		Create(&slabs).Error
+	if err != nil {
+		return fmt.Errorf("failed to create slabs %w", err)
+	}
+
+	// fetch the upserted slabs
+	for i := range slabs {
+		if err := tx.Raw("SELECT * FROM slabs WHERE `key` = ?", slabs[i].Key).
+			Scan(&slabs[i]).
+			Error; err != nil {
+			return fmt.Errorf("failed to fetch slab: %w", err)
+		} else if slabs[i].DBContractSetID != contractSetID {
+			return fmt.Errorf("slab already exists in another contract set %v != %v", slabs[i].DBContractSetID, contractSetID)
 		}
+	}
 
-		// Create Slice.
-		slice := dbSlice{
+	// build slices
+	dbSlices := make([]dbSlice, len(slices))
+	for i := range slices {
+		slab := slabs[i]
+		dbSlices[i] = dbSlice{
 			DBSlabID:          slab.ID,
 			DBObjectID:        objID,
 			ObjectIndex:       uint(i + 1),
 			DBMultipartPartID: multiPartID,
-			Offset:            ss.Offset,
-			Length:            ss.Length,
-		}
-		err = tx.Create(&slice).Error
-		if err != nil {
-			return fmt.Errorf("failed to create slice %v/%v: %w", i+1, len(slices), err)
+			Offset:            slices[i].Offset,
+			Length:            slices[i].Length,
 		}
+	}
+
+	// if there are no slices we are done
+	if len(dbSlices) == 0 {
+		return nil
+	}
+
+	// create slices
+	err = tx.Create(&dbSlices).Error
+	if err != nil {
+		return fmt.Errorf("failed to create slice %w", err)
+	}
 
-		for j, shard := range ss.Shards {
-			sector := dbSector{
+	// build sectors
+	var sectors []dbSector
+	for i, ss := range slices {
+		slab := slabs[i]
+		for j := range ss.Shards {
+			sectors = append(sectors, dbSector{
 				DBSlabID:   slab.ID,
 				SlabIndex:  j + 1,
-				LatestHost: publicKey(shard.LatestHost),
-				Root:       shard.Root[:],
-			}
-			// create sector if it doesn't exist yet
-			if err := createOrUpdateSector(tx, &sector); err != nil {
-				return fmt.Errorf("failed to create sector %v/%v: %w", j+1, len(ss.Shards), err)
-			}
-			// Add contract and host to join tables.
-			var associatedContracts []dbContract
+				LatestHost: publicKey(ss.Shards[j].LatestHost),
+				Root:       ss.Shards[j].Root[:],
+			})
+		}
+	}
+
+	// create sector that don't exist yet
+	if err := createOrUpdateSector(tx, sectors); err != nil {
+		return fmt.Errorf("failed to create sectors: %w", err)
+	}
+
+	// build contract <-> sector links
+	sectorIdx := 0
+	var contractSectors []dbContractSector
+	for _, ss := range slices {
+		for _, shard := range ss.Shards {
+			sector := sectors[sectorIdx]
 			for _, fcids := range shard.Contracts {
 				for _, fcid := range fcids {
 					if _, ok := contracts[fcid]; ok {
-						associatedContracts = append(associatedContracts, contracts[fcid])
+						contractSectors = append(contractSectors, dbContractSector{
```

Comment:
I was going to mention the deep nesting here but I see why it's a bit of an tough spot. We could build a link array in `// build sectors` ([][]types.FileContractID) so we'd only need to loop that and get the sector id using `sectors[i].ID` but maybe that's worse.

Reply:
Yeah it's a bit awkward since we can't do this before calling `upsertSectors` and merging loops decreases readability a lot. So I'd tolerate the nesting here to avoid introducing something worse.